hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsck.java

```diff
@@ -0,0 +1,421 @@
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    int scannedItems = 0;
+
+    final Path rootPath = rawFS.qualify(p);
+    S3AFileStatus root = null;
+    try {
+      root = (S3AFileStatus) rawFS.getFileStatus(rootPath);
+    } catch (AWSBadRequestException e) {
+      throw new IOException(e.getMessage());
+    }
+    final List<ComparePair> comparePairs = new ArrayList<>();
+    final Queue<S3AFileStatus> queue = new ArrayDeque<>();
+    queue.add(root);
+
+    while (!queue.isEmpty()) {
+      final S3AFileStatus currentDir = queue.poll();
+      scannedItems++;
+
+      final Path currentDirPath = currentDir.getPath();
+      List<FileStatus> s3DirListing = Arrays.asList(rawFS.listStatus(currentDirPath));
+
+      // DIRECTORIES
+      // Check directory authoritativeness consistency
+      compareAuthoritativeDirectoryFlag(comparePairs, currentDirPath, s3DirListing);
+      // Add all descendant directory to the queue
+      s3DirListing.stream().filter(pm -> pm.isDirectory())
+              .map(S3AFileStatus.class::cast)
+              .forEach(pm -> queue.add(pm));
+
+      // FILES
+      // check files for consistency
+      final List<S3AFileStatus> children = s3DirListing.stream()
+              .filter(status -> !status.isDirectory())
+              .map(S3AFileStatus.class::cast).collect(toList());
+      final List<ComparePair> compareResult =
+          compareS3DirToMs(currentDir, children).stream()
+              .filter(comparePair -> comparePair.containsViolation())
+              .collect(toList());
+      comparePairs.addAll(compareResult);
+      scannedItems += children.size();
+    }
+    stopwatch.stop();
+
+    // Create a handler and handle each violated pairs
+    S3GuardFsckViolationHandler handler =
+        new S3GuardFsckViolationHandler(rawFS, metadataStore);
+    comparePairs.forEach(handler::handle);
+
+    LOG.info("Total scan time: {}s", stopwatch.elapsed(TimeUnit.SECONDS));
+    LOG.info("Scanned entries: {}", scannedItems);
+
+    return comparePairs;
+  }
+
+  /**
+   * Compare the directory contents if the listing is authoritative
+   * @param comparePairs the list of compare pairs to add to if it contains a violation
+   * @param currentDirPath the current directory path
+   * @param s3DirListing the s3 directory listing to compare with
+   * @throws IOException
+   */
+  private void compareAuthoritativeDirectoryFlag(List<ComparePair> comparePairs, Path currentDirPath,
+                                                 List<FileStatus> s3DirListing) throws IOException {
+    final DirListingMetadata msDirListing =
+        metadataStore.listChildren(currentDirPath);
+    if (msDirListing != null && msDirListing.isAuthoritative()) {
+      ComparePair cP = new ComparePair(s3DirListing, msDirListing);
+
+      if (s3DirListing.size() != msDirListing.numEntries()) {
+        cP.violations.add(Violation.AUTHORITATIVE_DIRECTORY_CONTENT_MISMATCH);
```

Comment:
Having seen this, it's a hard error to interpret if there are many entries. Imagine if there were thousands and the S3 store had one too many? At the very least the user needs to give the counts on the MS and S3 side to help hint which side is actually short of data

Reply:
added to the AuthDirContentMismatch log part in the handler.

Reply:
`  /**
   * The violation handler when there's a directory listing content mismatch.
   */
  public static class AuthDirContentMismatch extends ViolationHandler {

    public AuthDirContentMismatch(S3GuardFsck.ComparePair comparePair) {
      super(comparePair);
    }

    @Override
    public String getError() {
      final String str = String.format(
          "The content of an authoritative directory listing does "
              + "not match the content of the S3 listing. S3: %s, MS: %s",
          Arrays.asList(getS3DirListing()), getMsDirListing().getListing());
      return str;
    }
  }`